File: scripts/generate-search-index.js
#!/usr/bin/env node

/**
 * Search Index Generator
 * Generates search-index.json from all markdown files
 */

const fs = require('fs');
const path = require('path');

const ROOT_DIR = path.join(__dirname, '..');
const MANIFEST_PATH = path.join(ROOT_DIR, 'manifest.json');
const OUTPUT_PATH = path.join(ROOT_DIR, 'search-index.json');

/**
 * Parse YAML front matter from markdown
 * @param {string} markdown - Markdown content
 * @returns {Object} Object with metadata and content
 */
function parseFrontMatter(markdown) {
  const frontMatterRegex = /^---\s*\n([\s\S]*?)\n---\s*\n([\s\S]*)$/;
  const match = markdown.match(frontMatterRegex);
  
  if (match) {
    const yaml = match[1];
    const content = match[2];
    const metadata = {};
    
    yaml.split('\n').forEach(line => {
      const trimmed = line.trim();
      if (!trimmed || trimmed.startsWith('#')) return;
      
      const colonIndex = trimmed.indexOf(':');
      if (colonIndex > 0) {
        const key = trimmed.substring(0, colonIndex).trim();
        let value = trimmed.substring(colonIndex + 1).trim();
        
        // Remove quotes if present
        if ((value.startsWith('"') && value.endsWith('"')) ||
            (value.startsWith("'") && value.endsWith("'"))) {
          value = value.slice(1, -1);
        }
        
        metadata[key] = value;
      }
    });
    
    return { metadata, content };
  }
  
  return { metadata: {}, content: markdown };
}

/**
 * Get base file path (without language suffix)
 * @param {string} filePath - File path with possible language suffix
 * @returns {string} Base file path
 */
function getBaseFile(filePath) {
  return filePath.replace(/ - AR\.md$/, '.md').replace(/ -AR\.md$/, '.md');
}

/**
 * Detect language from file path
 * @param {string} filePath - File path
 * @returns {string} Language code ('en' or 'ar')
 */
function detectLanguage(filePath) {
  if (filePath.includes(' - AR.md') || filePath.includes(' -AR.md')) {
    return 'ar';
  }
  return 'en';
}

/**
 * Extract title from markdown (H1 heading)
 * @param {string} markdown - Markdown content
 * @returns {string|null} Title or null
 */
function extractTitleFromH1(markdown) {
  const h1Match = markdown.match(/^#\s+(.+)$/m);
  return h1Match ? h1Match[1].trim() : null;
}

/**
 * Get title from filename
 * @param {string} filePath - File path
 * @returns {string} Title from filename
 */
function getTitleFromFilename(filePath) {
  const filename = filePath.split('/').pop();
  let title = filename.replace(/\.md$/, '').replace(/\s*-\s*AR$/, '');
  return title;
}

/**
 * Process a markdown file and extract searchable content
 * @param {string} filePath - Path to markdown file
 * @param {Object} manifestEntry - Entry from manifest.json
 * @returns {Object|null} Search index entry or null if file doesn't exist
 */
function processMarkdownFile(filePath, manifestEntry) {
  const fullPath = path.join(ROOT_DIR, filePath);
  
  if (!fs.existsSync(fullPath)) {
    console.warn(`File not found: ${filePath}`);
    return null;
  }
  
  try {
    const content = fs.readFileSync(fullPath, 'utf8');
    const { metadata, content: markdownContent } = parseFrontMatter(content);
    
    // Get title: front matter > manifest > H1 > filename
    let title = metadata.title || 
                manifestEntry?.title || 
                extractTitleFromH1(markdownContent) || 
                getTitleFromFilename(filePath);
    
    // Get base path (without language suffix)
    const basePath = getBaseFile(filePath);
    
    // Detect language
    const language = detectLanguage(filePath);
    
    // Get group from manifest or infer from path
    const group = manifestEntry?.group || 
                  (filePath.startsWith('Apps/') ? 'apps' : 
                   filePath.startsWith('Technologies/') ? 'technologies' : 'root');
    
    return {
      path: basePath,
      title: title,
      content: markdownContent.trim(),
      language: language,
      group: group
    };
  } catch (error) {
    console.error(`Error processing ${filePath}:`, error.message);
    return null;
  }
}

/**
 * Generate search index
 */
function generateSearchIndex() {
  console.log('Generating search index...');
  
  // Load manifest
  let manifest = { files: [] };
  if (fs.existsSync(MANIFEST_PATH)) {
    try {
      const manifestContent = fs.readFileSync(MANIFEST_PATH, 'utf8');
      manifest = JSON.parse(manifestContent);
    } catch (error) {
      console.error('Error reading manifest.json:', error.message);
      process.exit(1);
    }
  } else {
    console.warn('manifest.json not found, scanning filesystem...');
  }
  
  // Create a map of file paths to manifest entries
  const manifestMap = new Map();
  manifest.files.forEach(file => {
    manifestMap.set(file.path, file);
  });
  
  // Process all files from manifest
  const searchIndex = [];
  let processedCount = 0;
  let errorCount = 0;
  
  manifest.files.forEach(file => {
    const entry = processMarkdownFile(file.path, file);
    if (entry) {
      searchIndex.push(entry);
      processedCount++;
    } else {
      errorCount++;
    }
  });
  
  // Write search index
  try {
    fs.writeFileSync(OUTPUT_PATH, JSON.stringify(searchIndex, null, 2), 'utf8');
    console.log(`\n✓ Search index generated successfully!`);
    console.log(`  - Processed: ${processedCount} files`);
    console.log(`  - Errors: ${errorCount} files`);
    console.log(`  - Output: ${OUTPUT_PATH}`);
  } catch (error) {
    console.error('Error writing search-index.json:', error.message);
    process.exit(1);
  }
}

// Run generator
if (require.main === module) {
  generateSearchIndex();
}

module.exports = { generateSearchIndex };

File: scripts/generate_search_index.py
import os
import json
import re


def parse_front_matter(content):
    """
    Parse YAML front matter from markdown content.
    Returns metadata dict and cleaned content.
    """
    front_matter_regex = re.compile(r"^---\s*\n(.*?)\n---\s*\n", re.DOTALL)
    match = front_matter_regex.match(content)

    metadata = {}
    clean_content = content

    if match:
        front_matter = match.group(1)
        clean_content = content[match.end() :]

        # Simple YAML parsing for title
        for line in front_matter.split("\n"):
            if ":" in line:
                key, value = line.split(":", 1)
                metadata[key.strip()] = value.strip().strip("\"'")

    return metadata, clean_content


def extract_title_from_content(content):
    """Try to find h1 header in content"""
    match = re.search(r"^#\s+(.+)$", content, re.MULTILINE)
    if match:
        return match.group(1).strip()
    return None


def get_file_group(path):
    """Determine group based on directory (fallback)"""
    if path.startswith("content/apps/"):
        return "apps"
    elif path.startswith("content/technologies/"):
        return "technologies"
    elif path.startswith("content/resources/"):
        return "resources"
    elif path.startswith("docs/"):
        return "docs"
    return "root"


def is_arabic_file(path):
    """Check if a file is an Arabic version based on directory path (fallback)"""
    # docs/ar/...  or  content/<category>/ar/...
    if path.startswith("docs/ar/"):
        return True
    if path.startswith("content/") and "/ar/" in path:
        return True
    return False


def generate_assets():
    search_data = []
    manifest_files = []

    target_dirs = ["docs", "content"]

    for root_dir in target_dirs:
        for dirpath, dirnames, filenames in os.walk(root_dir):
            if "images" in dirnames:
                dirnames.remove("images")

            for filename in filenames:
                if not filename.endswith(".md"):
                    continue

                filepath = os.path.join(dirpath, filename)
                # Need to strip the leading ./ if it were there, but since we walk 'docs' it is already clean
                rel_path = filepath.replace(os.sep, "/")

                try:
                    with open(filepath, "r", encoding="utf-8") as f:
                        content = f.read()

                    metadata, clean_content = parse_front_matter(content)

                    # Determine title: Front matter > H1 > Filename
                    title = metadata.get("title")
                    if not title:
                        title = extract_title_from_content(content)
                    if not title:
                        title = os.path.splitext(filename)[0]

                    # Determine language from front matter vs directory structure
                    language = metadata.get("language")
                    if not language:
                        is_ar = is_arabic_file(rel_path)
                        language = "ar" if is_ar else "en"

                    group = metadata.get("group")
                    if not group:
                        group = get_file_group(rel_path)

                    has_ar = False
                    if language == "en":
                        # Simplistic check if an arabic version exists by looking for it
                        ar_path = ""
                        if rel_path.startswith("docs/"):
                            ar_path = rel_path.replace("docs/", "docs/ar/")
                        elif rel_path.startswith("content/"):
                            parts = rel_path.split("/")
                            if len(parts) >= 3:
                                parts.insert(2, "ar")
                                ar_path = "/".join(parts)
                        if ar_path and os.path.exists(ar_path):
                            has_ar = True
                    else:
                        has_ar = True  # Arabic files count as having an arabic version inherently for manifest purposes

                    # Search Index Entry
                    search_data.append(
                        {
                            "path": rel_path,
                            "title": title,
                            "content": clean_content.strip(),
                            "language": language,
                            "group": group,
                        }
                    )

                    # Manifest Entry
                    manifest_entry = {
                        "path": rel_path,
                        "title": title,
                        "group": group,
                        "hasAR": has_ar,
                    }
                    if metadata.get("arTitle"):
                        manifest_entry["arTitle"] = metadata.get("arTitle")

                    if language != "ar":
                        manifest_files.append(manifest_entry)

                    print(f"Processed: {rel_path} ({language})")

                except Exception as e:
                    print(f"Error processing {rel_path}: {e}")

    # Sort the lists by path before writing to ensure deterministic output
    search_data.sort(key=lambda x: x["path"])
    manifest_files.sort(key=lambda x: x["path"])

    # Write to search-index.json
    with open("search-index.json", "w", encoding="utf-8") as f:
        json.dump(search_data, f, ensure_ascii=False, indent=2)

    print(f"Successfully generated search-index.json with {len(search_data)} entries.")

    # Write to manifest.json
    manifest_data = {"files": manifest_files}
    with open("manifest.json", "w", encoding="utf-8") as f:
        json.dump(manifest_data, f, ensure_ascii=False, indent=2)
        f.write("\n")
    print(f"Successfully generated manifest.json with {len(manifest_files)} entries.")


if __name__ == "__main__":
    generate_assets()
